x/metadata/spec/examples/kotlin/src/main/kotlin/Bech32.kt

```diff
@@ -0,0 +1,169 @@
+        private const val MIN_VALID_CODEPOINT = 33
+        private const val MAX_VALID_CODEPOINT = 126
+
+        private const val charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+        private val gen = intArrayOf(0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3)
+
+        /** Decodes a Bech32 String */
+        fun decode(bech32: String): Bech32Data {
+            require(bech32.length in MIN_VALID_LENGTH..MAX_VALID_LENGTH) { "invalid bech32 string length" }
+            require(bech32.toCharArray().none { c -> c.toInt() < MIN_VALID_CODEPOINT || c.toInt() > MAX_VALID_CODEPOINT })
+            { "invalid character in bech32: ${bech32.toCharArray().map { c -> c.toInt() }
+                .filter { c -> c < MIN_VALID_CODEPOINT || c > MAX_VALID_CODEPOINT }}" }
+
+            require(bech32 == bech32.toLowerCase() || bech32 == bech32.toUpperCase())
+            { "bech32 must be either all upper or lower case" }
+            require(bech32.substring(1).dropLast(CHECKSUM_SIZE).contains('1')) { "invalid index of '1'" }
+
+            val hrp = bech32.substringBeforeLast('1').toLowerCase()
+            val dataString = bech32.substringAfterLast('1').toLowerCase()
+
+            require(dataString.toCharArray().all { c -> charset.contains(c) }) { "invalid data encoding character in bech32"}
+
+            val dataBytes = dataString.map { c -> charset.indexOf(c).toByte() }.toByteArray()
+            val checkBytes = dataString.takeLast(CHECKSUM_SIZE).map { c -> charset.indexOf(c).toByte() }.toByteArray()
+
+            val actualSum = checksum(hrp, dataBytes.dropLast(CHECKSUM_SIZE).toTypedArray())
+            require(1 == polymod(expandHrp(hrp).plus(dataBytes.map { d -> d.toInt() }))) { "checksum failed: $checkBytes != $actualSum" }
+
+            return Bech32Data(hrp, convertBits(dataBytes.dropLast(CHECKSUM_SIZE).toByteArray(), 5, 8, false))
+        }
+
+        /**
+         * Encodes the provided hrp and data to a Bech32 address string.
+         * @param hrp the human readable portion (prefix) to use.
+         * @param eightBitData an array of 8-bit encoded bytes.
+         */
+        fun encode(hrp: String, eightBitData: ByteArray) =
+            encodeFiveBitData(hrp, convertBits(eightBitData, 8, 5, true))
+
+        /** Encodes 5-bit bytes (fiveBitData) with a given human readable portion (hrp) into a bech32 string. */
+        private fun encodeFiveBitData(hrp: String, fiveBitData: ByteArray): String {
+            return (fiveBitData.plus(checksum(hrp, fiveBitData.toTypedArray()))
+                .map { b -> charset[b.toInt()] }).joinToString("", hrp + "1")
+        }
+
+        /**
+         * ConvertBits regroups bytes with toBits set based on reading groups of bits as a continuous stream group by fromBits.
+         * This process is used to convert from base64 (from 8) to base32 (to 5) or the inverse.
+         */
+        private fun convertBits(data: ByteArray, fromBits: Int, toBits: Int, pad: Boolean): ByteArray {
```

Comment:
making this a private implementation detail of this class is a very good move 👍🏻